File: webapp/src/main/scala/org/linnando/mixemulator/webapp/LineOutputDeviceComponent.scala
package org.linnando.mixemulator.webapp

import angulate2.std._
import org.scalajs.dom.raw.{Blob, BlobPropertyBag, URL}
import rxjs.core.Subscription

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js
import scala.util.{Failure, Success}

@Component(
  selector = "line-output-device",
  templateUrl = "webapp/src/main/resources/line-output-device.component.html",
  styleUrls = @@@("webapp/src/main/resources/line-output-device.component.css")
)
class LineOutputDeviceComponent(virtualMachineService: VirtualMachineService) extends OnInit with OnDestroy {
  @Input()
  var deviceNum: Int = _
  var stateChangeSubscription: Subscription = _
  var lines: js.Array[String] = js.Array()

  override def ngOnInit(): Unit = {
    fetchDeviceData()
    stateChangeSubscription = virtualMachineService.stateChange.subscribe(_ => fetchDeviceData())
  }

  private def fetchDeviceData(): Unit = {
    virtualMachineService.lineDeviceData(deviceNum) onComplete {
      case Success(data) => lines = js.Array[String]() ++ data
      case Failure(e) => ErrorPopup.show(e)
    }
  }

  override def ngOnDestroy(): Unit = {
    stateChangeSubscription.unsubscribe()
  }

  def saveFile(): Unit = {
    val blobParts = lines.map(_ + "\n").asInstanceOf[js.Array[js.Any]]
    val options = js.Dynamic.literal("type" -> "text/plain").asInstanceOf[BlobPropertyBag]
    val blob = new Blob(blobParts, options)
    val url = URL.createObjectURL(blob)
    js.Dynamic.global.window.open(url)
  }
}

File: webapp/src/app/line-output-device/line-output-device.component.ts
import { Component, Input, OnDestroy, OnInit } from '@angular/core';
import { faDownload } from '@fortawesome/free-solid-svg-icons';
import { BsModalService } from 'ngx-bootstrap/modal';
import { Subscription } from 'rxjs';
import { ErrorPopupComponent } from '../error-popup/error-popup.component';
import { VirtualMachineService } from '../virtual-machine.service';

@Component({
  selector: 'mix-line-output-device',
  templateUrl: './line-output-device.component.html',
  styleUrls: ['./line-output-device.component.scss']
})
export class LineOutputDeviceComponent implements OnInit, OnDestroy {
  @Input() deviceNum = 0;
  stateChangeSubscription?: Subscription;
  lines: string[] = [];
  readonly faDownload = faDownload;

  constructor(private modalService: BsModalService, private virtualMachineService: VirtualMachineService) { }

  ngOnInit(): void {
    this.fetchDeviceData();
    this.stateChangeSubscription = this.virtualMachineService.stateChange.subscribe(_ => {
      this.fetchDeviceData();
    });
  }

  private async fetchDeviceData(): Promise<void> {
    try {
      this.lines = await this.virtualMachineService.lineDeviceData(this.deviceNum);
    } catch (e) {
      this.modalService.show(ErrorPopupComponent, { initialState: { exception: e } });
    }
  }

  ngOnDestroy(): void {
    this.stateChangeSubscription?.unsubscribe();
  }

  downloadFile(): void {
    const blobParts = this.lines.map(line => line + '\n');
    const blob = new Blob(blobParts, { 'type': 'text/plain' });
    const url = URL.createObjectURL(blob);
    window.open(url);
  }
}
